### DISCUSSION TITLE
Add Legend to Diagrams

### DISCUSSION THREAD

**Discussion by jamestiotio** [terrastruct/d2#847] [Feature Requests]
## Feature Request Description

I would like to request the capability of adding a legend to diagrams.

The use case that I have would be to highlight a "normal operating path" and other "fault handling" paths in a state machine, something along the lines of this:

![State Machine Diagram](https://user-images.githubusercontent.com/18364745/219933841-87cb0117-2700-4270-abc0-db84d5c97fc6.svg)

Searching online, we can also see various examples of engineering diagrams that highlight some kind of "active path" or "critical path" such as [this](https://www.researchgate.net/profile/Christophe-Muller-2/publication/220091717/figure/fig3/AS:649635798466671@1531896652759/RNVFF-active-path-during-normal-operating-mode.png) and [this](https://help.synthesisplatform.net/blocksim_reno11/images/phase_aircraft_phases.png) (credits to their original authors). This shows that this type of diagram is used quite prevalently and that it can benefit from this legend.

## Proposed Solution/Implementation

The legend feature should allow users to create a separate shape within the diagram that contains a key of elements and their corresponding meanings in text. Each element can be either a shape or a connection.

Users should be able to customize the appearance of each element and the legend box itself (e.g., fill color, border radius, etc.), just like any other shape. This would allow the users to categorize different elements depending on their style (e.g., opacity, fill, stroke, etc.) to give them their appropriate meanings.

## Benefits

- Improve the clarity and readability of complex diagrams that contain a lot of shapes, such as state machines.
- Allow users to better communicate the intended functionality of a system by including a legend embedded in the diagram itself.

## Alternatives

My current workaround to generate the SVG above would be to do something as shown [here](https://play.d2lang.com/?script=pFLNysJADLzvU8wLfB-e9-BRPOkzBBLKwpqWdIsH6btL-qMIBVe9lGbYmcwMeZQdYd5ICElTwd8eQ-dflnXm9qoRtwBMa-dfH9xURGMiGoAxjE-NNxwTXhnLtur3WRpR_p9bdeoCsPQlKc2JTq1dKOPciU1Ipfsq7QOlPJjgSMo5afOB642EfsIRxUj7jky0TOhyZBG7TWu_67wk-lbsHgAA__8%3D). Another example would be [this](https://play.d2lang.com/?script=gge2lUIRKES4uAwVdO0UjECEMYgwARGmVgrVXAoKYPsgTBAH5BorhfSi1NQ8LgWFWq5aLi5TuBZjuDGGeDQXpabAtOakpqfmpehBwgmkDyqQklpckpmXCHGjW35ReWJRCpGuwTQByVyIRVYKTonJ2fjNxHAkFg-B0qCVQklRYl5xQWJRal4JWBSaSqwUDLD6knJzUAKHXMMAAQAA__8%3D), which would generate the following diagram:

![Directional Diagram](https://user-images.githubusercontent.com/18364745/219933968-04287fd7-d4ef-4807-a32e-a98656897735.svg)

Basically, we add several invisible elements to act as the legend and style them appropriately:

```d2
legend.source -> legend.destination: Forward {
  style: {
    stroke: green
  }
}

legend.destination -> legend.source: Backward {
  style: {
    stroke: red
  }
}

legend: {
  style: {
    fill: transparent
    opacity: 0
  }
}

legend.source: {
  style: {
    fill: transparent
    opacity: 0
  }
}

legend.destination: {
  style: {
    fill: transparent
    opacity: 0
  }
}
```

However, this approach has its limitations:
- The layout engine will consider the legend container along with its children as a separate root-level shape, and thus, the positioning of the legend can be inconsistent, especially in more complex diagrams with more shapes. The legend container and its children cannot use the `near` keyword to help with its positioning relative to the whole diagram as `constant near keys cannot be set on shapes with children` and `constant near keys can only be set on root level shapes`. This is demonstrated by the 2 diagrams above, which puts the legend in different places. This can also be seen [here](https://play.d2lang.com/?script=gge2lUIRKES4uAwVdO0UjECEMYgwARGmVgrVXAoKYPsgTBAH5BorhfSi1NQ8LgWFWq5aLi5TuBZjuDGGeDQXpabAtOakpqfmpehBwgmkDyqQklpckpmXCHGjW35ReWJRCpGuwTQByVyIRVYKTonJ2fjNxHAkFg-B0qCVQklRYl5xQWJRal4JWBSaSqwUDLD6knJzUAKHXMMAAQAA__8%3D&layout=tala), where the `direction: right` directive leads to the TALA layout engine putting the legend to the right of the actual diagram.
- The `legend`, `source`, and `destination` texts are technically still in the diagram, which means that they can be found/extracted by carefully analyzing the SVG output. That said, they are not visible in the PNG output.

Alternatively, we can manually draw the legend using another tool such as [diagrams.net](https://app.diagrams.net/) and package that diagram together with the D2-generated diagram, such as by putting them together in a ZIP file or combining them using other image processing tools. However, this is a little clunky and it would be great if D2 can generate the legend by itself. Adding the declaration to generate the legend into the D2 source file itself would also help keep track of changes to the legend itself when users check in the D2 source file into version control.

**Comment by alixander** [comment]
Once D2 supports near keywords on containers, which isn't much work, the first limitation will be crossed off. (https://github.com/terrastruct/d2/issues/848)

The second limitation I don't agree with, unless there's something I'm not understanding. The legend would always be visible and in the outputs of both SVG and PNG. If you really want to hide the texts you can even set the labels to "".

The bigger question is whether we want to support a first-class way of defining legends, with custom keywords and all.

This task feels really large. I imagine the variations of legends people want will be a lot.
- Icons
- Colored shapes
- Mini styled arrow

Pretty endless. I think the best thing for us to do right now is to do the work needed to build a custom legend, and then put a section in docs like, "Recipes" or something that shows constructions of legends.

**Reply by jamestiotio** [reply]
Ah, I was not aware that you could set the label to `""`. Thank you for that! In that case, the second limitation should be a non-issue.

I don't mind using my workaround, assuming that the first limitation is eliminated. Once that's done, I can also help add a section in docs to show how to construct custom legends using this approach. If you have any suggestions on how to create legends in a better way, do feel free to bring them up!

Regarding the addition of defining first-class legends, I am not particularly for or against it. As long as there is a way to build custom legends, with D2 offering some level of support (and it does not have to be creating a native, first-class entity specifically for legends), then it should be fine.
